### PR TITLE
reuse algorithm.fill while building SkipTable

### PR DESCRIPTION
**Pull request by n5m**

### CODE REVIEW

File: lib/pure/strutils.nim [hunk 1]
```diff
@@ -1779,17 +1779,7 @@ func initSkipTable*(a: var SkipTable, sub: string) {.rtl,
...
-    a[chr(i + 0)] = m
-    a[chr(i + 1)] = m
-    a[chr(i + 2)] = m
-    a[chr(i + 3)] = m
-    a[chr(i + 4)] = m
-    a[chr(i + 5)] = m
-    a[chr(i + 6)] = m
-    a[chr(i + 7)] = m
-    i += 8
+  algorithm.fill[int](a, m)
```
Comment (timotheecour) [inline lib/pure/strutils.nim]:
```suggestion
  fill(a, m)
```

is loop unrolling (as before this PR) still happening with fill with optimizations on?
or is previous code a case of premature optimzation?

Comment (n5m) [reply lib/pure/strutils.nim]: When not using `algorithm.fill`, the generated C code looks like
```c
N_LIB_PRIVATE N_NIMCALL(void, nsuInitSkipTable)(NI* a, NimStringDesc* sub) {
    NI m;
    NI i;
    m = (sub ? sub->Sup.len : 0);
    i = ((NI) 0);
    {
        while (1) {
            if (!(i <= ((NI) 248))) goto LA2;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 0)))))))))- 0] = m;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 1)))))))))- 0] = m;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 2)))))))))- 0] = m;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 3)))))))))- 0] = m;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 4)))))))))- 0] = m;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 5)))))))))- 0] = m;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 6)))))))))- 0] = m;
            a[(((NU8)(((NIM_CHAR) (((NI) ((NI)(i + ((NI) 7)))))))))- 0] = m;
            i += ((NI) 8);
        } LA2: ;
    }
    /// rest of the method...
```

When using `algorithm.fill`, the generated C code looks like
```c
N_LIB_PRIVATE N_NIMCALL(void, fill_pureZstrutils_1383)(NI* a, NI aLen_0, NI value) {
    NI xX60gensym163_;
    xX60gensym163_ = ((NI) 0);
    {
        while (1) {
            if (!(xX60gensym163_ <= (aLen_0-1))) goto LA2;
            a[xX60gensym163_] = value;
            xX60gensym163_ += ((NI) 1);
        } LA2: ;
    }
}
N_LIB_PRIVATE N_NIMCALL(void, nsuInitSkipTable)(NI* a, NimStringDesc* sub) {
    NI m;
    m = (sub ? sub->Sup.len : 0);
    fill_pureZstrutils_1383(a, 256, m);
    // rest of the method...
```

So while `algorithm.fill` does not generate unrolled C code, I think a modern C compiler would be able to correctly optimize the code generated by `algorithm.fill`.

These were compiled with `-d:danger --gc:none --checks:off --opt:speed`.

Comment (timotheecour) [reply lib/pure/strutils.nim]: that's what i thought too (premature optimization); and in fact I'm now seeing memset can be 30x slower that a loop (when dealing with chars) (pending double checking that the loop doesn't get optimized away); can share details if needed (refs D20210603T142208)

Comment (n5m) [reply lib/pure/strutils.nim]: I am certainly curious about the details on that. What does the identifier "D20210603T142208" refer to?

**Comment by timotheecour** [comment]
@n5m 
> I am certainly curious about the details on that. What does the identifier "D20210603T142208" refer to?

here you go, this shows `memset` is 30X slower than a loop in this benchmark (12x slower if adding `--passc:-flto`); note that changing the numbers and the code lead to similar results. Please let me know if you find a flaw in this benchmark!

```nim
when true: # D20210603T142208
  import std/times
  proc c_memset*(p: pointer, value: cint, size: csize_t): pointer {.importc: "memset", header: "<string.h>", discardable.}
  template algo_loop(s, n, val, start): untyped =
    for i in start..<start+n:
      s[i] = val
  template algo_memset(s, n, val, start): untyped =
    c_memset(s[start].addr, val.ord.cint, n.csize_t)
  template mainAux(algo)=
    var s: string
    let n = 100
    let n2 = 50
    s.setLen n
    var c = 0
    let t = cpuTime()
    for i in 0..<100000:
      for j in 0..<n2:
        algo(s, n - n2, i.char, j) # similar conclusions with `algo(s, n - n2, (i+j).char, j)`

    for z in 0..<n: c+=cast[int](s[z])
      # prevent optimizing out
      # indending this to call this after each `i` would still make memset slower, but by less
    
    let t2 = cpuTime()
    echo (astToStr(algo), c, t2 - t)

  proc main()=
    for i in 0..<10:
      echo()
      mainAux(algo_loop)
      mainAux(algo_memset)
  main()
```

nim -v: 654a20166eabb614cbf0379c67115981f0532079
`nim r -d:danger main.nim`

prints:
("algo_loop", 15741, 0.0003969999999999998)
("algo_memset", 15741, 0.011760999999999999)

("algo_loop", 15741, 0.0003040000000000022)
("algo_memset", 15741, 0.011001999999999998)

("algo_loop", 15741, 0.00033099999999999796)
("algo_memset", 15741, 0.010647)
...


the question is whether there are benchmarks showing memset being faster than a loop

(in those experiments i'm using osx)

File: lib/pure/strutils.nim (hunk 1 above)
Comment (timotheecour) [reply lib/pure/strutils.nim]: => see https://github.com/nim-lang/Nim/pull/18138#issuecomment-854936534

**Pull request merged**